Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java

```diff
@@ -216,69 +267,25 @@ public void setInvocationCount(int newInvocationCount) {
 
-        // When the number of threads is 1, the overhead of thread pools must be avoided
-        if (numThreads == 1) {
-            // This logic is duplicated in the CoreDistanceRunnable nested class below
-            for (int point = 0; point < data.length; point++) {
-                // Sorted nearest distances found so far
-                double[] kNNDistances = new double[numNeighbors];
-                Arrays.fill(kNNDistances, Double.MAX_VALUE);
-
-                for (int neighbor = 0; neighbor < data.length; neighbor++) {
-                    if (point == neighbor) {
-                        continue;
-                    }
-                    double distance = getDistance(data[point], data[neighbor], distanceType);
-
-                    // Check at which position in the nearest distances the current distance would fit.
-                    // k is typically small, but if cases with larger values of k become prevalent, this should be replaced
-                    // with a binary search
-                    int neighborIndex = numNeighbors;
-                    while (neighborIndex >= 1 && distance < kNNDistances[neighborIndex - 1]) {
-                        neighborIndex--;
-                    }
-
-                    // Shift elements in the array to make room for the current distance
-                    // The for loop could be written as an arraycopy, but the result is not particularly readable, and
-                    // numNeighbors is typically quite small
-                    if (neighborIndex < numNeighbors) {
-                        for (int shiftIndex = numNeighbors - 1; shiftIndex > neighborIndex; shiftIndex--) {
-                            kNNDistances[shiftIndex] = kNNDistances[shiftIndex - 1];
-                        }
-                        kNNDistances[neighborIndex] = distance;
-                    }
-                }
-                // The core distance for the point is the distance to the furthest away neighbor
-                coreDistances.set(point, kNNDistances[numNeighbors - 1]);
-            }
-        } else { // This makes the core distance calculations with multiple threads
-            ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
-            for (int point = 0; point < data.length; point++) {
-                executorService.execute(new CoreDistanceRunnable(data, numNeighbors, distanceType, point, coreDistances));
-            }
-            executorService.shutdown();
-            try {
-                boolean finished = executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.MINUTES);
-                if (!finished) {
-                    throw new RuntimeException("Parallel execution failed");
-                }
-            } catch (InterruptedException e) {
-                throw new RuntimeException("Parallel execution failed", e);
-            }
+        NeighboursBruteForceFactory nbff = new NeighboursBruteForceFactory(distType, numThreads);
```

Comment:
Maybe this should be pushed into a configurable field on this trainer? That way it would be easy to benchmark the different implementations.

Reply:
Again, I agree that we'll certainly want the `NeighboursQueryFactory` to be field here, it's just a matter of when to add it.

Reply:
I'm fine with making this change later, provided we make it before the next release (which is still many months away).

Reply:
Yes, the `NeighboursKDTreeFactory` and `NeighboursKDTree` classes which will leverage these discussed changes should be ready well in advance of this timeframe.